src/client/dfuse/pil4dfs/dfs_dcache.c

```diff
@@ -0,0 +1,663 @@
+	if (rec == NULL)
+		D_GOTO(error, rc = -DER_NOMEM);
+
+	rec->dr_obj = obj;
+	atomic_init(&rec->dr_ref, 0);
+	atomic_init(&rec->dr_deleted, false);
+	memcpy(&rec->dr_key_child_prefix[0], &dcache->dd_key_root_prefix[0], DCACHE_KEY_PREF_SIZE);
+	memcpy(&rec->dr_key[0], &dcache->dd_key_root_prefix[0], DCACHE_KEY_PREF_SIZE);
+	rec->dr_key_len = DCACHE_KEY_PREF_SIZE - 1;
+	rc = d_hash_rec_insert(&dcache->dd_dir_hash, rec->dr_key, rec->dr_key_len, &rec->dr_entry,
+			       true);
+	if (rc == 0)
+		D_GOTO(out, rc = -DER_SUCCESS);
+
+error:
+	D_FREE(rec);
+out:
+	return rc;
+}
+
+static int
+dcache_create_act(dfs_t *dfs, uint32_t bits, uint32_t rec_timeout, dfs_dcache_t **dcache)
+{
+	dfs_dcache_t *dcache_tmp = NULL;
+	dfs_obj_t    *obj        = NULL;
+	daos_obj_id_t obj_id;
+	mode_t        mode;
+	int           rc;
+
+	D_ALLOC_PTR(dcache_tmp);
+	if (dcache_tmp == NULL)
+		D_GOTO(error, rc = -DER_NOMEM);
+
+	dcache_tmp->dd_dfs = dfs;
+
+	rc = dfs_lookup(dfs, "/", O_RDWR, &obj, &mode, NULL);
+	if (rc != 0)
+		D_GOTO(error, rc = daos_errno2der(rc));
+	rc = dfs_obj2id(obj, &obj_id);
+	D_ASSERT(rc == 0);
+	rc = snprintf(&dcache_tmp->dd_key_root_prefix[0], DCACHE_KEY_PREF_SIZE,
+		      "%016" PRIx64 "-%016" PRIx64 ":", obj_id.hi, obj_id.lo);
+	D_ASSERT(rc == DCACHE_KEY_PREF_SIZE - 1);
+
+	rc = d_hash_table_create_inplace(D_HASH_FT_MUTEX | D_HASH_FT_LRU, bits, NULL,
+					 &dcache_hash_ops, &dcache_tmp->dd_dir_hash);
+	if (rc != 0)
+		D_GOTO(error, rc);
+
+	dcache_tmp->destroy_fn     = dcache_destroy_act;
+	dcache_tmp->find_insert_fn = dcache_find_insert_act;
+	dcache_tmp->drec_incref_fn = drec_incref_act;
+	dcache_tmp->drec_decref_fn = drec_decref_act;
+	dcache_tmp->drec_del_at_fn = drec_del_at_act;
+	dcache_tmp->drec_del_fn    = drec_del_act;
+
+	rc = dcache_add_root(dcache_tmp, obj);
+	if (rc != 0)
+		D_GOTO(error, rc);
+
+	*dcache = dcache_tmp;
+	D_GOTO(out, rc = -DER_SUCCESS);
+
+error:
+	dfs_release(obj);
+	D_FREE(dcache_tmp);
+out:
+	return rc;
+}
+
+static int
+dcache_destroy_act(dfs_dcache_t *dcache)
+{
+	d_list_t *rlink = NULL;
+	int       rc;
+
+	while ((rlink = d_hash_rec_first(&dcache->dd_dir_hash)) != NULL) {
+		bool deleted;
+
+		D_ASSERT(atomic_load(&dlist2drec(rlink)->dr_ref) == 1);
+		deleted = d_hash_rec_delete_at(&dcache->dd_dir_hash, rlink);
+		D_ASSERT(deleted);
+	}
+
+	rc = d_hash_table_destroy_inplace(&dcache->dd_dir_hash, false);
+	if (rc != 0) {
+		DL_ERROR(rc, "d_hash_table_destroy_inplace() failed");
+		D_GOTO(out, rc);
+	}
+
+	D_FREE(dcache);
+
+out:
+	return rc;
+}
+
+static inline dcache_rec_t *
+dcache_get(dfs_dcache_t *dcache, const char *key, size_t key_len)
+{
+	d_list_t *rlink;
+
+	D_ASSERT(dcache != NULL);
+	D_ASSERT(key != NULL);
+	D_ASSERT(key_len > 0);
+
+	rlink = d_hash_rec_find(&dcache->dd_dir_hash, key, key_len);
+	if (rlink == NULL)
+		return NULL;
+	return dlist2drec(rlink);
+}
+
+static inline int
+dcache_add(dfs_dcache_t *dcache, dcache_rec_t *parent, const char *name, const char *key,
```

Comment:
How does the key differ from the name here?  Shouldn't we by using the `directory entry` as the key, otherwise how can we search for it?

Reply:
The key is constructed in the following way:
<low bit dirobj_id>-<high bit dirobj_id>:<dir name>

Reply:
We spoke about this off-line, the key is already formatted for the previous (failed) lookup so the key is passed in here in addition to the name as it's already been generated, but at this point yes, the key is the oid+name as you say.  I was confused by the string seemingly being passed in twice as two different parameters.